### ISSUE TITLE
research: DataAssimBench — solver robustness guards, TLM/Lyapunov validation utilities, and a published 4DVar benchmark

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/vardax#53]
# Comparative Analysis: `DataAssimBench` vs `vardax`

## Context

[StevePny/DataAssimBench](https://github.com/StevePny/DataAssimBench) (`dabench`) is a JAX-based data assimilation benchmarking suite: synthetic-truth generators (Lorenz63/96, SQG turbulence, QG variants), an observation generator, and four DA "cyclers" — 3D-Var (CG), ETKF, traditional 4D-Var (TLM + BiCGSTAB), and **backprop-4D-Var** (autodiff gradient descent through the model, optax-driven). Its companion repo [DataAssimBench-Examples](https://github.com/StevePny/DataAssimBench-Examples) contains the full, published benchmark protocol from Solvik, Penny & Hoyer (2024) comparing backprop-4DVar against traditional 4DVar.

This survey asked what vardax can borrow. Architecturally, nothing: `dabench`'s flag-based `DACycler` inheritance (`_in_4d`, `_uses_ensemble`) is exactly the vertical hierarchy that vardax's D14 "seven horizontal peers" design rejects, and its dual `StateVector`/xarray representation is what vardax's JAX-array-only core avoids. What *is* worth taking: (1) battle-tested **robustness guards** for gradient-descent inner loops; (2) an **explicit tangent-linear-model (TLM) integrator** as a *validation* tool for vardax's autodiff adjoints; (3) a **Lyapunov-exponent utility** for principled window-length choices in demos/docs; (4) a ready-made, citable **benchmark protocol** for Epic 12 and a proven **tutorial ladder** for Epic 11.

---

### A. Backprop-4DVar robustness guards (`dabench/dacycler/_var4d_backprop.py`)

Their gradient-descent 4DVar (the method closest to FourDVarNet's inner loop and to a gradient-descent `StrongFourDVar`) ships with practical safeguards:

```python
Var4DBackprop(
    system_dim, delta_t, model_obj,
    learning_rate=0.5,
    lr_decay=0.5,            # exponential LR decay via optax schedule
    num_iters=3,
    steps_per_window=None,
    loss_growth_limit=10,    # divergence guard
)
```

- **Divergence guard**: if the inner-loop loss exceeds `loss_growth_limit ×` its starting value, the loop aborts/flags instead of silently producing garbage — on chaotic systems a too-large step routinely sends J(x₀) to ∞.
- **NaN detection** inside the optimization loop, same rationale.
- **Exponential learning-rate decay** through an optax schedule (cheap, matters a lot for convergence on L96-class problems).
- A sibling variant with **Hessian preconditioning** (`_var4d_backprop_exacthessian.py`) — approximate/exact Hessian applied as a preconditioner on the gradient step.

### B. Explicit TLM integration (`dabench/data/_data.py`)

`Data.generate(..., return_tlm=True)` integrates the augmented ODE alongside the state:

```text
dx/dt = f(x)
dM/dt = J(x)·M ,   M(0) = I        # J = ∂f/∂x; M(t) is the exact propagator
```

yielding the tangent-linear propagator `M(tₖ)` at every output step (shape `(time, d, d)`). Their traditional 4DVar consumes these for the inner-loop normal equations `(B⁻¹ + ∑ Mᵀ Hᵀ R⁻¹ H M) δx = b`.

### C. Lyapunov exponents (`dabench/data/_data.py`)

`calc_lyapunov_exponents_series()` / `_final()` — Eckmann et al. (1985) algorithm: propagate the TLM, QR-re-orthonormalize at rescale intervals, average log diagonal of R; includes convergence checks. Self-contained, ~100 lines.

### D. Published benchmark protocol (examples repo)

`examples/da_cycler/Solvik_Penny_Hoyer_2024-Backprop4DVar/` — a complete, citable experiment design:

- **Stage 1 (RayTune)**: hyperparameter search over learning rate + decay per configuration.
- **Stage 2 (full run)**: backprop-4DVar vs. traditional 4DVar vs. **no-DA baseline**, on validation and test segments.
- **Axes**: L96 system dimension ∈ {6, 20, 36, 72, 144, 256}; observed-variable count ∈ {6 … 36}; obs error σ ∈ {0.1 … 2.0}; plus PyQG-JAX and QGS testbeds.
- **Outputs**: analysis RMSE *and* wall-clock time, reported jointly; figure notebooks per testbed.

The examples repo overall follows a clean tutorial ladder: data generators → observers → DA cyclers → paper reproduction.

---

### A. Already in vardax (direct equivalents)

| dabench feature | vardax equivalent | Path | Notes |
|---|---|---|---|
| `Var3D` (CG) | `ThreeDVar` (optimistix GaussNewton/BFGS/NonlinearCG) | `src/vardax/_src/models/threedvar.py` | ours: pluggable minimisers, posterior adapters |
| `Var4D` (TLM + BiCGSTAB) | `IncrementalFourDVar` (GN outer + CG inner + CVT via gaussx) | `src/vardax/_src/models/incremental_fourdvar.py` | ours is the proper incremental formulation |
| `Var4DBackprop` | `StrongFourDVar` (diffrax/optimistix adjoints), `FourDVarNet` inner loop | `src/vardax/_src/models/strong_fourdvar.py`, `src/vardax/_src/model.py` | but see B1 — they have guards we lack |
| L63/L96 generators | `simulate_lorenz96` + L63 helpers | `src/vardax/_src/utils/dynamical_systems.py` | ours lack TLM/Lyapunov (C1, C2) |
| B, R, H as arrays | `lineax.AbstractLinearOperator` + gaussx structure | throughout | we're ahead (structured, matrix-free) |

#### B1. Inner-loop robustness guards (HIGH PRIORITY)
- **`dabench`**: divergence guard (`loss_growth_limit`), NaN detection, exponential LR decay, optional Hessian preconditioning — see §1.A.
- **vardax**: the FourDVarNet inner loop (`src/vardax/_src/solver.py` — unrolled / one-step / implicit variants) and the gradient-descent paths in the classical methods run unguarded: a diverging inner loop currently surfaces only as bad output or NaNs downstream.
- **What's needed**: a small `GuardConfig` (or kwargs) on the solver loops in `_src/solver.py` and on `StrongFourDVar`'s minimiser slot: track `loss₀`, stop (and record in the `Posterior.provenance` dict) when `loss > limit × loss₀` or `isnan(loss)`; expose an optax LR-decay schedule as a documented default in `_src/training.py` / `fit_demo()`.
- **Impact**: stability on chaotic systems; honest provenance (a flagged divergence beats a silently wrong analysis); near-zero cost.

#### C1. Explicit TLM integrator as an adjoint-validation utility (HIGH PRIORITY)
- **What it is**: §1.B — integrate `dM/dt = J(x)·M` alongside the trajectory for L63/L96, giving exact propagators `M(tₖ)`.
- **Why useful**: vardax rightly takes adjoints from diffrax/optimistix rather than hand-coding TLMs — but that means nothing currently *validates* those adjoints against ground truth. With explicit `M`, tests can assert `jax.vjp` through the rollout ≡ `Mᵀ` (and `jvp` ≡ `M`) to tolerance — a stronger calibration check than the linear-Gaussian agreement suite (`tests/test_linear_gaussian_agreement.py`), which only checks the *fixed point*, not the *gradient path*. Also satisfies the `pipekit_cycle.TangentLinearModel` protocol for free on the demo systems.
- **Where in vardax**: `src/vardax/_src/utils/dynamical_systems.py` (e.g. `simulate_with_tlm(rhs, jac, x0, ...)` via `diffrax` on the augmented system); new test module `tests/test_adjoint_vs_tlm.py`.

#### C2. Lyapunov-exponent utility (MEDIUM PRIORITY)
- **What it is**: §1.C — Eckmann '85 with QR re-orthonormalization; depends on C1's TLM.
- **Why useful**: the leading Lyapunov exponent λ₁ gives a principled assimilation-window choice (window ≲ 1/λ₁ for strong-constraint 4DVar) instead of magic numbers in the L63/L96 notebooks; also a quantitative hook for the math-reference docs and for choosing `steps_per_window` in benchmarks.
- **Where in vardax**: `src/vardax/_src/utils/dynamical_systems.py`; demonstrated in the existing L63/L96 notebooks.

#### C3. Adopt the Solvik/Penny/Hoyer protocol for Epic 12 benchmarking (HIGH PRIORITY)
- **What it is**: §1.D — a published two-stage benchmark (hyperparameter search, then RMSE + wall-clock vs. traditional 4DVar and a no-DA baseline) across system size, obs density, and obs error.
- **Why useful**: Epic 12 (performance benchmarking, `docs/design/boundaries.md`) is currently unspecified. Adopting this protocol wholesale means the design work is done *and* the numbers are externally comparable: a thin adapter exposing a vardax `AnalysisStep` as a `dabench.dacycler.DACycler` would let `StrongFourDVar` / `IncrementalFourDVar` / `FourDVarNet` run inside DataAssimBench itself and be checked against the paper's reference results.
- **Where in vardax**: benchmark scripts/notebooks under Epic 12; optional `dabench` adapter in a `contrib/` or docs-only module (keep `dabench` out of core deps).

#### C4. Tutorial ladder for Epic 11 (MEDIUM PRIORITY, docs only)
- **What it is**: the examples repo's progression — data generation → observation sampling → single method end-to-end → method comparison → paper reproduction.
- **Why useful**: Epic 11 (real-world tutorials) can copy this structure; vardax's existing notebooks (`notebooks/01…08`) already cover the middle rungs but lack the explicit observation-sampling rung and a reproduction capstone.

### D. vardax has it, `dabench` doesn't (context only — no action)
- Seven horizontal `AnalysisStep` peers vs. their flag-based inheritance; posterior adapters (`Posterior` + Laplace/GN-Hessian/Ensemble) — they return point estimates only; structured covariances via gaussx/lineax vs. their dense B = I default; weak-constraint and incremental 4DVar; amortized inference; upstream adjoints (diffrax/optimistix) vs. their hand-rolled loops.
- Their `xarray_jax` threading of `xr.Dataset` through `lax.scan` is a useful **negative data point** for vardax's open coordax question (`docs/design/boundaries.md` open question 1): the dual-representation friction visible throughout `dabench` supports keeping vardax JAX-array-only and doing coordinates at the pipekit/xrtoolz boundary.

---

## 3. Summary Table

| Feature | `dabench` | `vardax` | Status |
|---|---|---|---|
| 3DVar / 4DVar / incremental | ✓ (basic) | ✓ (richer) | We're ahead |
| Gradient-descent 4DVar | ✓ | ✓ | **Already have** |
| Divergence/NaN guards + LR decay in inner loop | ✓ | ✗ | **Enhancement needed** (B1) |
| Hessian-preconditioned gradient step | ✓ | ✗ (GN path exists) | Optional, fold into B1 |
| Explicit TLM integration (augmented ODE) | ✓ | ✗ | **Missing** (C1) |
| Adjoint-vs-TLM validation tests | ✗ (implicit) | ✗ | **Missing** (C1) |
| Lyapunov exponents | ✓ | ✗ | **Missing** (C2) |
| Published benchmark protocol + baselines | ✓ | ✗ (Epic 12 unspecified) | **Missing** (C3) |
| Posterior covariance outputs | ✗ | ✓ | We're ahead |
| Structured operators (gaussx/lineax) | ✗ | ✓ | We're ahead |

---

### Phase 1: Cheap wins
1. B1 — robustness guards in `_src/solver.py` + LR-decay default in training utilities.

### Phase 2: Validation foundations
2. C1 — `simulate_with_tlm` + `tests/test_adjoint_vs_tlm.py`.
3. C2 — Lyapunov exponents (depends on C1); wire λ₁-based window choice into the L63/L96 notebooks.

### Phase 3: Benchmarking (Epic 12)
4. C3 — adopt Solvik/Penny/Hoyer protocol; optional `dabench` `DACycler` adapter for external validation.

### Phase 4: Docs (Epic 11)
5. C4 — tutorial ladder structure.

---

## 5. Proposed Follow-up Issues

- [ ] `feat(solver): divergence/NaN guards + LR-decay defaults for inner loops` — covers Phase 1 / B1
- [ ] `feat(utils): explicit TLM integration for L63/L96 + adjoint-vs-TLM validation tests` — covers Phase 2 / C1
- [ ] `feat(utils): Lyapunov-exponent computation (Eckmann '85) + window-length guidance in notebooks` — covers Phase 2 / C2
- [ ] `feat(bench): Epic 12 benchmark via Solvik/Penny/Hoyer protocol (+ optional dabench adapter)` — covers Phase 3 / C3
- [ ] `docs: Epic 11 tutorial ladder modeled on DataAssimBench-Examples` — covers Phase 4 / C4

---

## 6. Key Synergies

1. **C1 → C2 → C3 form a chain**: the TLM utility powers both the Lyapunov computation and a principled window-length choice for the Epic 12 benchmarks, so the benchmark configs stop being magic numbers.
2. **B1 + posterior provenance**: recording divergence-guard trips in `Posterior.provenance` turns a numerical safeguard into auditable metadata — consistent with vardax's D10 posterior-adapter philosophy and unavailable in `dabench` itself.
3. **C3 closes the credibility loop**: running vardax methods *inside* a third-party published benchmark is the strongest external validation available short of a paper.

---

## References
- DataAssimBench — https://github.com/StevePny/DataAssimBench
- DataAssimBench-Examples — https://github.com/StevePny/DataAssimBench-Examples
- Solvik, Penny & Hoyer (2024) — backprop-4DVar benchmark (notebooks under `examples/da_cycler/Solvik_Penny_Hoyer_2024-Backprop4DVar/`)
- Eckmann & Ruelle (1985) — Lyapunov-spectrum algorithm used by `calc_lyapunov_exponents_series`

## Relationships
- Parent (theme epic, if any): —
- Blocked by: —
- Blocks (follow-up issues from §5 reference this as parent research): TBD
- Related: jejjohnson/pipekit#44 (Observer + OSSE benchmark recipe); jejjohnson/xrtoolz#244 (state-vector round-trip + observation-sampling masks)
